### PR TITLE
Remaining value evaluation inc. read/write functions

### PR DESCRIPTION
**Pull request by cjllanwarne**
* Wiring in the remaining draft 3 engine functions' value evaluators
* Modified value evaluation to keep a running track of its side-effect file generation
* Also allowed value evaluation to run value mapping if used in placeholder interpolation

Note: red thumb required because there's some changes to WomIdentifier undoing some of the changes in the forkjoin PR and making them again, but correcter this time. 

This PR ended up with far wider-ranging changes than I originally expected (even requiring changes in draft 2 and CWL) so if you see something that smells wrong... ⚠️ ❗️ ❗️ ⚠️ 

~~NB: this PR will go down by about 700 lines once the `forkjoin` PR merges~~ DONE!

### CODE REVIEW

**Comment by cjllanwarne** [comment]
Will return...

**Comment by cjllanwarne** [comment]
Still looking a bit iffy. Watch out for a comeback...

File: wdl/transforms/draft3/src/main/scala/wdl/draft3/transforms/wdlom2wom/WorkflowDefinitionElementToWomWorkflowDefinition.scala [hunk 1]
```diff
@@ -61,7 +61,7 @@ object WorkflowDefinitionElementToWomWorkflowDefinition {
 
     def graphNodeCreationFold(currentValidation: ErrorOr[List[GraphNode]], next: WorkflowGraphElement): ErrorOr[List[GraphNode]] = {
       def outputName(node: GraphNode, port: OutputPort) = node match {
-        case _: CallNode => port.identifier.workflowLocalName
+//        case _: CallNode => port.identifier.localName.value
```
Comment (kshakir) [inline wdl/transforms/draft3/src/main/scala/wdl/draft3/transforms/wdlom2wom/WorkflowDefinitionElementToWomWorkflowDefinition.scala]: Remove this dead code or add more comments as to why it's sticking around.

File: wom/src/main/scala/wom/graph/CallNode.scala [hunk 2]
```diff
@@ -96,9 +98,9 @@ object TaskCall {
     // The FQN combines the name of the task to the name of the input or output
     def identifier(name: LocalName) = WomIdentifier(name, taskDefinitionLocalName.combineToFullyQualifiedName(name))
 
-    def linkOutput(call: GraphNode)(output: OutputDefinition): ErrorOr[GraphNode] = call.outputByName(output.name).map(out => PortBasedGraphOutputNode(
-      identifier(output.localName), output.womType, out
-    ))
+//    def linkOutput(call: GraphNode)(output: OutputDefinition): ErrorOr[GraphNode] = call.outputByName(output.name).map(out => PortBasedGraphOutputNode(
+//      identifier(output.localName), output.womType, out
+//    ))
```
Comment (kshakir) [inline wom/src/main/scala/wom/graph/CallNode.scala]: Delete this or add more comments about the situation for when it should return.

File: wom/src/main/scala/wom/graph/CallNode.scala [hunk 3]
```diff
@@ -122,13 +124,18 @@ object TaskCall {
     val uniqueIdentifier = WomIdentifier(taskDefinition.name)
     val callWithInputs = callNodeBuilder.build(uniqueIdentifier, taskDefinition, inputDefinitionFold)
 
-    for {
-      outputs <- taskDefinition.outputs.traverse(linkOutput(callWithInputs.node) _)
+    val outputNodes: ErrorOr[Seq[GraphOutputNode]] = callWithInputs.node.outputPorts.map(output => PortBasedGraphOutputNode(
+      identifier(LocalName(output.internalName)), output.womType, output
+    )).toList.validNel
+    val result = for {
+      outputs <- outputNodes // taskDefinition.outputs.traverse(linkOutput(callWithInputs.node) _)
```
Comment (kshakir) [inline wom/src/main/scala/wom/graph/CallNode.scala]: What does this comment mean? Is it an "aka"?

File: wom/src/main/scala/wom/graph/GraphNodePort.scala [hunk 4]
```diff
@@ -53,9 +58,9 @@ object GraphNodePort {
   }
   case class GraphNodeOutputPort(override val identifier: WomIdentifier, womType: WomType, graphNode: GraphNode) extends OutputPort
   object ExpressionBasedOutputPort {
-    def apply(localName: LocalName, womType: WomType, graphNode: GraphNode, expression: WomExpression): ExpressionBasedOutputPort = {
-      ExpressionBasedOutputPort(WomIdentifier(localName, graphNode.identifier.fullyQualifiedName.combine(localName.value)), womType, graphNode, expression)
-    }
+//    def apply(outputIdentifier: WomIdentifier, womType: WomType, graphNode: GraphNode, expression: WomExpression): ExpressionBasedOutputPort = {
+//      new ExpressionBasedOutputPort(outputIdentifier, womType, graphNode, expression)
+//    }
```
Comment (kshakir) [inline wom/src/main/scala/wom/graph/GraphNodePort.scala]: Delete or add more comments.

**Comment by cjllanwarne** [comment]
Overriding pullapprove - looks like 2 thumbs to me

**Comment by kshakir** [comment]
Here's the pullapprove audit: https://pullapprove.com/broadinstitute/cromwell/pull-request/3505/reviews/

**Pull request merged**